src/LinkAdder.php

```diff
@@ -0,0 +1,115 @@
+<?php
+
+namespace Spatie\Crawler;
+
+use Tree\Node\Node;
+use GuzzleHttp\Psr7\Uri;
+use InvalidArgumentException;
+use Psr\Http\Message\UriInterface;
+use Symfony\Component\DomCrawler\Link;
+use Symfony\Component\DomCrawler\Crawler as DomCrawler;
+
+class LinkAdder
+{
+    /** @var \Spatie\Crawler\Crawler */
+    protected $crawler;
+
+    public function __construct(Crawler $crawler)
+    {
+        $this->crawler = $crawler;
+    }
+
+    public function addFromHtml(string $html, UriInterface $foundOnUrl)
+    {
+        $allLinks = $this->extractLinksFromHtml($html, $foundOnUrl);
+
+        collect($allLinks)
+            ->filter(function (UriInterface $url) {
+                return $this->hasCrawlableScheme($url);
+            })
+            ->map(function (UriInterface $url) {
+                return $this->normalizeUrl($url);
+            })
+            ->each(function (UriInterface $url) use ($foundOnUrl) {
+                $node = $this->crawler->addToDepthTree($url, $foundOnUrl);
+
+                if (strpos($url->getPath(), '/tel:') === 0) {
+                    return;
+                }
+
+                if (! $this->shouldCrawl($node)) {
+                    return;
+                }
+
+                if ($this->crawler->maximumCrawlCountReached()) {
+                    return;
+                }
+
+                $crawlUrl = CrawlUrl::create($url, $foundOnUrl);
+
+                $this->crawler->addToCrawlQueue($crawlUrl);
+            });
+    }
+
+    /**
+     * @param string                         $html
+     * @param \Psr\Http\Message\UriInterface $foundOnUrl
+     *
+     * @return \Illuminate\Support\Collection|\Tightenco\Collect\Support\Collection|null
+     */
+    protected function extractLinksFromHtml(string $html, UriInterface $foundOnUrl)
```

Comment:
Can we use a nullable typehint here?

Reply:
I don't think so, because either `Illuminate\Support\Collection` or `Tightenco\Collect\Support\Collection` is returned.